x-pack/legacy/plugins/ml/server/models/job_service/new_job/categorization.ts

```diff
@@ -0,0 +1,298 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { ML_RESULTS_INDEX_PATTERN } from '../../../../common/constants/index_patterns';
+import { CATEGORY_EXAMPLES_MULTIPLIER } from '../../../../common/constants/new_job';
+import { CategoryId, Category, Token } from '../../../../common/types/categories';
+
+type callWithRequestType = (action: string, params: any) => Promise<any>;
+
+export function categorizationExamplesProvider(callWithRequest: callWithRequestType) {
+  async function categorizationExamples(
+    indexPatternTitle: string,
+    query: any,
+    size: number,
+    categorizationFieldName: string,
+    timeField: string | undefined,
+    start: number,
+    end: number,
+    analyzer?: any
+  ) {
+    if (timeField !== undefined) {
+      const range = {
+        range: {
+          [timeField]: {
+            gte: start,
+            format: 'epoch_millis',
+          },
+        },
+      };
+
+      if (query.bool === undefined) {
+        query.bool = {};
+      }
+      if (query.bool.filter === undefined) {
+        query.bool.filter = range;
+      } else {
+        if (Array.isArray(query.bool.filter)) {
+          query.bool.filter.push(range);
+        } else {
+          query.bool.filter.range = range;
+        }
+      }
+    }
+
+    const results = await callWithRequest('search', {
+      index: indexPatternTitle,
+      size,
+      body: {
+        _source: categorizationFieldName,
+        query,
+      },
+    });
+    const examples: string[] = results.hits?.hits
+      ?.map((doc: any) => doc._source[categorizationFieldName])
+      .filter((example: string | undefined) => example !== undefined);
+
+    let tokens: Token[] = [];
+    try {
+      const { tokens: tempTokens } = await callWithRequest('indices.analyze', {
+        body: {
+          ...getAnalyzer(analyzer),
+          text: examples,
+        },
+      });
+      tokens = tempTokens;
+    } catch (error) {
+      // fail silently, the tokens could not be loaded
+      // an empty list of tokens will be returned for each example
+    }
+
+    const lengths = examples.map(e => e.length);
+    const sumLengths = lengths.map((s => (a: number) => (s += a))(0));
+
+    const tokensPerExample: Token[][] = examples.map(e => []);
+
+    const { tokenizer } = getAnalyzer(analyzer); // remove this once ml_classic bug is fixed
+
+    tokens.forEach((t, i) => {
+      for (let g = 0; g < sumLengths.length; g++) {
+        const extra = tokenizer === 'ml_classic' ? 0 : g; // remove this once ml_classic bug is fixed
```

Comment:
elastic/elasticsearch#50655 was merged about 6 hours ago.  It's probably best to delay merging this PR until the ml_classic fix has propagated into a unified snapshot.  If this PR is merged with the workaround then the QA team will probably find the functionality to be broken by the time they test it.